DO NOT MERGE: json file format example

tvanfossen opened the pull request:
A few example json files that I've been using

By no means do I believe this is an end-all-be-all! its purely been a rapid prototyping tool to build/add new views, thought it might be useful for conversation on this topic

Have an accompanying view manager/builder that will control transitions between views, trigger relevant animations,etc.  Application specific to my device, but hopefully helps drive conversation.

Some nomenclature was shortened for json size.

"aos" == "AnimationOnStart" (on view enter/creation)
"aoe" == "AnimationOnExit" (on view exit/teardown
"adc" == "AnimationOnDataChange" (100ms event)

tvanfossen-bissell committed:
example schema and json views

tvanfossen commented:
@kisvegabor 

More complete json example, cleaning.json showing how bloated a view can look under this format, but hopefully can help drive some ideas.

Not tied to a specific format by any means, this same style could be easily adapted to yaml/xml/css

tvanfossen commented:
Each file represents 1 view

Each view is giving a list of next possible views, each of with specific triggers in order to reach that screen.

Each view gives a list of widgets on that screen

Each widget contains a set of animations matched to some view state

Each animation contains relevant triggers, property to animate upon, values/time to animate over etc

kisvegabor commented:
Thanks @tvanfossen!

These JSON files both "describing LVGL" and "markup language for the UI" too.

For the "describing LVGL" part I also meant something like your [view_schema.json](https://github.com/lvgl/markup_temporary/pull/1/files#diff-89d90d0210c5001943818458206f3c7a525e06c8bff2a540b943adce199a830d). 

As discussed in the original issue I suggest focusing on "describing LVGL" first.

tvanfossen-bissell committed:
update schema

tvanfossen-bissell committed:
updated yaml example

tvanfossen-bissell committed:
example hello tinyfsm state

tvanfossen commented:
Updated with latest schema in use in private project. This schema is a bit more concise and links back to relevant properties in LVGL

Switched from JSON to YAML for view files. Fonts/text styles and language strings handled in separate yaml and are generated at build time using custom script + lv_font_conv. Assets are generated from PNG files at build time using lv_img_conv

For every view yaml in a project (such as the example hello.yaml), at build time a script converts the yaml structure into a tinyfsm state machine. View transitions/updates/animations are then handled by corresponding view state.

Generation script is still WIP and not included here, but I can broach the topic to pushing it upstream if its of interest. Highly tied to use of tinyfsm and our unified data model implementation, but has resulted in very clean "low code" UI dev.

Where the original commit relied on loading a json file from spiffs at runtime and needed to be size conscious, everything is now built at compile time with the generation script and runs wonderfully with minimal overhead. Generation script also accommodates placement and animation of lottie files

![image](https://user-images.githubusercontent.com/17345424/169860238-d4859931-feef-47a8-8b13-3da22a995b6c.png)

In this example, DURABLE_MAIN_STATEOFCHARGE is a unique key generated at build time by our unified data model repo  to represent the state of charge value coming from our device. When the unique key is changed, it sends out a tinyfsm event that the view *can* react to and run animations, change views, accordingly.

Our application has yet to call for touch, and as such there is a large gap in the schema around touch to be worked on at a future time.

kisvegabor commented:
Hi,

Thanks again for sharing! YAML feel much more readable to me. 

I guess proprties like this maps to local styles:
https://github.com/lvgl/markup_temporary/blob/7878f9a65825b9a81798b62aca295ad147d5dc6c/hello.yaml#L27

Do you have a view on how to describe widget parts and states?

tvanfossen commented:
At the moment, widget parts/states are inferred by the property of the yaml - more to "get it working" than "get it perfectly mapping to lvgl parts/states/styles" so theres still certainly large gaps 

![image](https://user-images.githubusercontent.com/17345424/170063555-e47f1d1b-fa7e-4482-ad97-8d7fa0bc1f5d.png)

So for the above widget in the yaml, the generation script in turn produces:

![image](https://user-images.githubusercontent.com/17345424/170063775-f319b824-03b3-46e4-9942-7fb47e0364d8.png)

At entry into this particular state, it builds that arc widget as widget_0, and performs any entry animations for that widget.

![image](https://user-images.githubusercontent.com/17345424/170064037-30f5f8bb-4b39-4945-9d50-efdff1d98e92.png)

For reference, the animations started here are what get built from the aos array in the previous post

tvanfossen commented:
the set_*some_anim_property* calls are utility functions that have been stubbed out to make the generation of animations for various properties of different widgets

kisvegabor commented:
Got it, thanks!
